### ISSUE TITLE
Could not cast Dip.WeakBox<Any> to ... if using .weakSingleton in separate module

### ISSUE THREAD

**Issue by sc0rch** [AliSoftware/Dip#128]
@ilyapuchka 
I found another frustrating issue, here the module definition:
```
let editUserEmailModule = DependencyContainer() { container in
  // EditUserEmailWireframe
  let editUserEmailWireframe = container.register(.weakSingleton) {
    EditUserEmailWireframe()
      as EditUserEmailWireframe
  }
  
  // EditUserEmailInteractor
  let editUserEmailInteractor = container.register(.weakSingleton) {
    EditUserEmailInteractor()
      as EditUserEmailInteractorInput
  }
  
  // ConfirmUserEmailInteractor
  let confirmUserEmailInteractor = container.register(.weakSingleton) {
    ConfirmUserEmailInteractor()
      as ConfirmUserEmailInteractorInput
  }
  
  // EditUserEmailViewPresenter
  let editUserEmailViewPresenter = container.register(.weakSingleton) {
    EditUserEmailViewPresenter()
      as EditUserEmailViewPresenter
  }
  container.register(editUserEmailViewPresenter, type: EditUserEmailModuleInterface.self)
  container.register(editUserEmailViewPresenter, type: EditUserEmailInteractorOutput.self)
  container.register(editUserEmailViewPresenter, type: ConfirmUserEmailInteractorOutput.self)
  
  // EditUserEmailViewController
  let editUserEmailViewController = container.register(.weakSingleton) {
    EditUserEmailViewController(nibName: "EditUserEmailViewController", bundle: nil)
      as EditUserEmailViewController
  }
  container.register(editUserEmailViewController, type: EditUserEmailViewInterface.self)

  // ConfirmUserEmailViewController
  let confirmUserEmailViewController = container.register(.weakSingleton) {
    ConfirmUserEmailViewController(nibName: "ConfirmUserEmailViewController", bundle: nil)
      as ConfirmUserEmailViewController
  }
  container.register(confirmUserEmailViewController, type: ConfirmUserEmailViewInterface.self)
}

let dipRoot = DependencyContainer() { root in
  ...

  mainModule.collaborate(
    with: root, editUserEmailModule
  )
  editUserEmailModule.collaborate(
    with: root, mainModule
  )
}

// Later in code
let controller = try? editUserEmailModule.resolve() as EditUserEmailViewController
```

I'm receiving the error:
```
Reusing previously resolved instance Dip.WeakBox<Any>
Could not cast value of type 'Dip.WeakBox<Any>' (0x1223b31d0) to 'MobitrackClientApp.EditUserEmailViewController' (0x10bd79020).
```

If I'll change ``.weakSingleton`` to ``.shared``, crash be gone, but code will start to recreate some modules instead of reusing. Maybe I don't understand clearly how collaboration and scopes works together.

<img width="720" alt="dip_could_not_cast" src="https://cloud.githubusercontent.com/assets/2544587/19233760/41d4e768-8ef0-11e6-9c66-832bb9d46ba3.png">

**Comment by ilyapuchka** [comment]
What version of Dip do you use?

**Comment by sc0rch** [comment]
Dip 5.0.2 version.

**Comment by sc0rch** [comment]
There are probably some issues in our code with order of weak & strong initializations after DI integration. The refactoring process for big projects can be real pain.
Sorry for bothering you.

**Comment by ilyapuchka** [comment]
Not a problem at all, it's good when someone faces non-trivial issues. 
I guess there is a bug in reusing previously resolved weak singleton instances when they are being released.

Most likely what happens is that on this line
`case .weakSingleton: return (weakSingletons[key] as? WeakBoxType)?.unboxed ?? weakSingletons[key]` there is `WeakBox<Any>` in the `weakSingletons` collection. But it's wrapped value is `nil`, because it is stored in a box by weak reference and looks like no one outside container holds strong reference to it any more. So `??` detects `nil` on the left side and return just right side `weakSingletons[key]` which is a `WeakBox<Any>`.

There is clearly an issue in the container, but also you may have some unneded weak reference to this controller instead of strong in your code.

**Comment by sc0rch** [comment]
@ilyapuchka 
We are creating some modules dynamically.
Maybe some of weak singletons were released in our code too early. We are investigating it.

**Comment by ilyapuchka** [comment]
@sc0rch can you please test your code agains #129 ? Should fix the crash, though you might still have a wrong behaviour of creating new instances of controller instead of reusing, what would mean that it previously resolved instance was already released.

**Comment by sc0rch** [comment]
@ilyapuchka 
Fix #129 works. No crash anymore. ViewController and it dependencies have been initialized. Other our issues are belong to scopes and modules, we will fix them ourself eventually.

**Comment by ilyapuchka** [comment]
Feel free to ask for help and thanks for helping to find the issue.
